### ISSUE TITLE
Problem with stackalloc code analyzer inside loop

### ISSUE THREAD

**Issue by mehmet-soyturk-qonic** [dotnet/roslyn-analyzers#7122]
### Description

Possibly related to issue [#33782](https://github.com/dotnet/runtime/issues/33782).

There is a warning for stackalloc statements inside loops, but only if the stackalloc references a type.

There is a warning for this (good):
```
for (int i = 0; i < 5; i++)
{
    Span<int> x = stackalloc int[] { 5, 10 };
}
```

But there is no warning for this (bad):
```
for (int i = 0; i < 5; i++)
{
    Span<int> x = stackalloc[] { 5, 10 };
}
```

There are also unnecessary warnings for expressions in loops that are executed only once:
```
foreach (int n in stackalloc int[] { 1, 3 })
{
}
```
```
for (int i = stackalloc int[] { 1, 3 }.Length; i < 7; i++)
{
     Console.WriteLine($"Hello {i}");
}
```

I'm using Visual Studio 2022.

### Reproduction Steps

Try this code:

```
class StackallocTest
{
    public static void Main()
    {
        for (int i = 0; i < 5; i++)
        {
            Span<int> x = stackalloc int[] { 5, 10 };
        }

        for (int i = 0; i < 5; i++)
        {
            Span<int> x = stackalloc[] { 5, 10 };
        }

        foreach (int n in stackalloc int[] { 1, 3 })
        {
        }

        for (int i = stackalloc int[] { 1, 3 }.Length; i < 7; i++)
        {
        }
    }
}
```

### Expected behavior

There are no warnings if stackalloc would be executed just once.

There are warnings even if stackalloc does not contain a type argument.

### Actual behavior

There are warnings for stackalloc that is executed just once for loop initialization.

There are no warnings if stackalloc does not contain a type argument.

**Comment by mehmet-soyturk-qonic** [comment]
Probably.

Microsoft Visual Studio Community 2022 Version 17.3.6
VisualStudio.17.Release/17.3.6+32929.385
Microsoft .NET Framework Version 4.8.09032
C# Tools   4.3.0-3.22470.13+80a8ce8d5fdb9ceda4101e2acb8e8eb7be4ebcea (don't know whether that is relevant)

I suppose half of this bug report is a duplicate then.

The other half was about the lack of warning with stackalloc with type inference. This might still be there:
```
for (int i = 0; i < 5; i++)
{
    Span<int> x = stackalloc[] { 5, 10 };
}
```

**Comment by stephentoub** [comment]
> There is no warning when stackalloc does not directly reference a type

Yes, that should be addressed. Thanks.
